### PR TITLE
Fix bugs in DynamicBrokerSelector

### PR DESCRIPTION
**Pull request by vrajat**
DynamicBrokerSelector.getBrokers returned a list of zkServers instead of brokers.

BrokerSelectorUtils.getTablesCommonBrokers made inplace updates to lists in brokerData. Make a copy of the lists so that the contents of brokerData is unchanged.

Member variables of DynamicBrokerSelector have been changed to protected scope so that the class can be derived.

tags: bugfix

### CODE REVIEW

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/utils/BrokerSelectorUtils.java [hunk 1]
```diff
@@ -57,11 +59,13 @@ public static List<String> getTablesCommonBrokers(List<String> tableNames, Map<S
...
-      commonBrokers.retainAll(tablesBrokersList.get(i));
-    }
-    return commonBrokers;
+    // Make a copy of the brokersList for the first table. retainAll does inplace modifications corrupting brokerData.
+    Set<String> commonBrokers = tablesBrokersList.stream()
+        .skip(1) // skip because the HashSet is initialized with the first list.
+        .collect(() -> new HashSet<>(tablesBrokersList.get(0)), Set::retainAll, Set::retainAll);
+    List<String> commonBrokerList = new ArrayList<>(commonBrokers.size());
+    commonBrokerList.addAll(commonBrokers);
+    return commonBrokerList;
```
Comment (gortiz) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/utils/BrokerSelectorUtils.java:68]: I found this quite more difficult to read (and probably more expensive in allocation) than the original. Why not just something like:

```
    List<String> commonBrokers = new ArrayList<>(tablesBrokersList.get(0));
    for (int i = 1; i < tablesBrokersList.size(); i++) {
      commonBrokers.retainAll(tablesBrokersList.get(i));
    }
    return commonBrokers;
```

Alternatively commonBrokers can be changed to a set if you want to optimize that computation (at the cost of an extra allocation)

Comment (vrajat) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/utils/BrokerSelectorUtils.java]: Done. In terms of efficiency, Set vs List doesnt matter as the size of the list is small. So moved back to list.

**Comment by yashmayya** [comment]
@vrajat looks like there's an NPE in `org.apache.pinot.client.ConnectionFactoryTest` with these changes:

```
WARN [ExternalViewReader] [main] Exception while reading External view from zookeeper
java.lang.NullPointerException: Cannot load from byte/boolean array because "znodeContents" is null
	at org.apache.pinot.client.ExternalViewReader.unpackZnodeIfNecessary(ExternalViewReader.java:177) ~[classes/:?]
	at org.apache.pinot.client.ExternalViewReader.getTableToBrokersMap(ExternalViewReader.java:144) ~[classes/:?]
	at org.apache.pinot.client.DynamicBrokerSelector.refresh(DynamicBrokerSelector.java:83) ~[classes/:?]
	at org.apache.pinot.client.DynamicBrokerSelector.<init>(DynamicBrokerSelector.java:61) ~[classes/:?]
	at org.apache.pinot.client.DynamicBrokerSelector.<init>(DynamicBrokerSelector.java:64) ~[classes/:?]
	at org.apache.pinot.client.ConnectionFactoryTest$1.<init>(ConnectionFactoryTest.java:40) ~[test-classes/:?]
	at org.apache.pinot.client.ConnectionFactoryTest.testZkConnection(ConnectionFactoryTest.java:40) ~[test-classes/:?]

```

**Review by yashmayya — COMMENTED** [review]
@vrajat looks like there's an NPE in `org.apache.pinot.client.ConnectionFactoryTest` with these changes:

```
WARN [ExternalViewReader] [main] Exception while reading External view from zookeeper
java.lang.NullPointerException: Cannot load from byte/boolean array because "znodeContents" is null
	at org.apache.pinot.client.ExternalViewReader.unpackZnodeIfNecessary(ExternalViewReader.java:177) ~[classes/:?]
	at org.apache.pinot.client.ExternalViewReader.getTableToBrokersMap(ExternalViewReader.java:144) ~[classes/:?]
	at org.apache.pinot.client.DynamicBrokerSelector.refresh(DynamicBrokerSelector.java:83) ~[classes/:?]
	at org.apache.pinot.client.DynamicBrokerSelector.<init>(DynamicBrokerSelector.java:61) ~[classes/:?]
	at org.apache.pinot.client.DynamicBrokerSelector.<init>(DynamicBrokerSelector.java:64) ~[classes/:?]
	at org.apache.pinot.client.ConnectionFactoryTest$1.<init>(ConnectionFactoryTest.java:40) ~[test-classes/:?]
	at org.apache.pinot.client.ConnectionFactoryTest.testZkConnection(ConnectionFactoryTest.java:40) ~[test-classes/:?]

```

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ public String selectBroker(String... tableNames) {
 
   @Override
   public List<String> getBrokers() {
-    return _brokerList;
+    return _allBrokerListRef.get();
```
Comment (soumitra-st) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java:115]: Was there a consumer of this method that relied on the list of ZK servers, as returned previously?

Comment (soumitra-st) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java:115]: Shall we return an immutable list of brokers?

**Review by soumitra-st — APPROVED** [review]
LGTM

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java (hunk 2 above)
Comment (vrajat) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java:115]: > Was there a consumer of this method that relied on the list of ZK servers, as returned previously?
There is no consumer in the Pinot code base. I dont know if external clients use it. I verified with @Jackie-Jiang that this is a bug. 

> Shall we return an immutable list of brokers?
I am keeping the PR as small as possible.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java [hunk 3]
```diff
@@ -45,11 +45,11 @@ public class DynamicBrokerSelector implements BrokerSelector, IZkDataListener {
...
-  private final AtomicReference<Map<String, List<String>>> _tableToBrokerListMapRef = new AtomicReference<>();
-  private final AtomicReference<List<String>> _allBrokerListRef = new AtomicReference<>();
-  private final ZkClient _zkClient;
-  private final ExternalViewReader _evReader;
-  private final List<String> _brokerList;
+  protected final AtomicReference<Map<String, List<String>>> _tableToBrokerListMapRef = new AtomicReference<>();
+  protected final AtomicReference<List<String>> _allBrokerListRef = new AtomicReference<>();
+  protected final ZkClient _zkClient;
+  protected final ExternalViewReader _evReader;
+  protected final List<String> _brokerList;
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java:52]: Do we need to keep this list? This is very confusing as it contains the ZK servers

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java [hunk 4]
```diff
@@ -45,11 +45,11 @@ public class DynamicBrokerSelector implements BrokerSelector, IZkDataListener {
   private static final Logger LOGGER = LoggerFactory.getLogger(DynamicBrokerSelector.class);
   private static final Random RANDOM = new Random();
 
-  private final AtomicReference<Map<String, List<String>>> _tableToBrokerListMapRef = new AtomicReference<>();
-  private final AtomicReference<List<String>> _allBrokerListRef = new AtomicReference<>();
-  private final ZkClient _zkClient;
-  private final ExternalViewReader _evReader;
-  private final List<String> _brokerList;
+  protected final AtomicReference<Map<String, List<String>>> _tableToBrokerListMapRef = new AtomicReference<>();
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java:48]: Any specific reason you want to make them `protected`? Are you going to extend this class?

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java (hunk 4 above)
Comment (vrajat) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java:48]: Yes. I want to have the option of extending this class.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java (hunk 3 above)
Comment (vrajat) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/DynamicBrokerSelector.java]: Removed.

**Pull request merged**